### PR TITLE
Allocate target's tuple table slot in PortalHeapMemory during split partition

### PR DESCRIPTION
**Pull request by armenatzoglou**
When we execute an ALTER TABLE statement that splits default partition into a "New" and a new Default partition, we perform the following 5 steps:
* copy the tuples located in the default partition into a temporary table,
* drop default partition,
* create the “New“ partition,
* create the new DEFAULT partition, and
* copy all tuples from the temporary table into the newly created partitions.

The last step is executed in split_rows, where for each tuple in the temporary table, 
* we determine the target partition/table, 
* construct the slot of the target table (if is NULL), and 
* store the tuple into the target table. 

Currently, we allocate the tuple table slot of the target table into a per-tuple memory context. If the size of that memory context exceeds 50KB, then we reset it. This will cause issues, since it will free target table's slot and at the next iteration (copy the next tuple) we will try to access the attributes of a freed (and not NULL) slot.
In addition, the target table slot lifespan is much longer than an individual tuple. So it is not correct to allocate the slot in a per-tuple memory context. 

To solve this issue, we allocate target's tuple table slot in PortalHeapMemory. This is the CurrentMemoryContext that is used in split_rows before we start copying tuples. PortalHeapMemory is used already for storing the tuple table slot of the temporary table and the ResultRelInfo of the target tables. PortalHeapMemory is not freed while we copy tuples. After copying all tuples to the new partitions, we drop target tables' slots.  

@foyzur @karthijrk Please take a look if you have some time.

### CODE REVIEW

File: src/backend/commands/tablecmds.c [hunk 1]
```diff
@@ -15248,6 +15248,10 @@ split_rows(Relation intoa, Relation intob, Relation temprel)
 	rrib->ri_partSlot = MakeSingleTupleTableSlot(RelationGetDescr(intob));
 	map_part_attrs(temprel, intoa, &rria->ri_partInsertMap, true);
 	map_part_attrs(temprel, intob, &rrib->ri_partInsertMap, true);
+	Assert(NULL != rria->ri_RelationDesc);
```
Comment (foyzur) [inline src/backend/commands/tablecmds.c]: What is the performance implication of always creating these slots? If I remember correctly, creating those was optional, and was only necessary at the edge cases.

**Comment by foyzur** [comment]
LGTM

**Comment by danielgustafsson** [comment]
Seems reasonable.

File: src/backend/commands/tablecmds.c (hunk 1 above)
Comment (gcaragea) [inline src/backend/commands/tablecmds.c]: It's negligible. We only create them once for the entire "ALTER TABLE" command, so it really doesn't matter. Thanks.

**Comment by armenatzoglou** [comment]
Pushed (c0e1f00c2532d1e2ef8d3b409dc8fee901a7cfe2). Thanks a lot for the review.

**Pull request closed without merging**